### PR TITLE
Shared library/executable testing in CI and update to the Geant4 RPATH

### PR DESCRIPTION
**Pull request by pshriwise**
## Description
This PR adds shared library testing to our CI. The CI tests will fail at first for reasons described in #673, but once the update to `build_geant4.sh` is applied we should see them succeed. I'll wait to update the docker image until someone has had a look and give me the :+1: to do so.

To enable the shared build testing, I've moved a lot of the actual building of DAGMC into `travis.tests.sh`, so our files 

One other small change I've added is the `CTEST_OUTPUT_ON_FAILURE` flag to provide more information in the CI log if a test does fail.

## Motivation and Context
We should be testing both static and shared builds of DAGMC on CI. We have plenty of spare time and resources on Travis.

### CODE REVIEW

**Comment by pshriwise** [comment]
If there are no objections, I'm going to go head and update the CI Docker image tomorrow evening with the changes in `build_geant4.sh` from this PR.

**Comment by gonuke** [comment]
Do you know why it hasn't even tried to run tests on this branch?

**Comment by gonuke** [comment]
I think this should probably be coordinated with #668 - I'm not sure which should come first?

**Comment by pshriwise** [comment]
@gonuke It looks like they did run, but the results aren't appearing here for some reason. :thinking: 

https://travis-ci.org/github/svalinn/DAGMC/builds/679251870

I'll chat with @bam241 tomorrow about how best to proceed with our two CI-related PRs.

**Comment by pshriwise** [comment]
@bam241 and I settled on getting #668 in first. I'll update this one on top of those changes and update CI when appropriate.

In the meantime I'll see what I can do about getting the CI test results to appear here.

**Comment by pshriwise** [comment]
Thanks again for tackling this @bam241!

**Comment by gonuke** [comment]
This looks OK to me. We'll need to push a new docker image once this is merged.

**Comment by bam241** [comment]
with this change you basically remove everything form the `install.sh` and move it into the `test.sh`
which almost remove the need of the `instal`l step in our `travis` workflow.

I am really not sure this is the best solution but:
Would it be possible to have both static and shared coexist in the same container, and just swap environment variable, so installs stays in the `install.sh` script and test in the `test.sh` ?

**Comment by bam241** [comment]
> with this change you basically remove everything form the `install.sh` and move it into the `test.sh`
> which almost remove the need of the `instal`l step in our `travis` workflow.
> 
> I am really not sure this is the best solution but:
> Would it be possible to have both static and shared coexist in the same container, and just swap environment variable, so installs stays in the `install.sh` script and test in the `test.sh` ?

we could also separate STATIC/SHARED build in our travis jobs, but I feel like we already have too many...

**Comment by pshriwise** [comment]
Fair point @bam241. I've updated the `install.sh` script to install and test both the static and shared versions of DAGMC in different locations. Was this about what you had in mind?

**Comment by bam241** [comment]
yes it was an option.
we could had more granularity in the travis building process also... 
have job for static and job for shared (if they don't need to coexist....)

**Comment by pshriwise** [comment]
Very true. I guess I agreed with your sentiment that we have enough CI jobs as it is :)

**Comment by bam241** [comment]
we lost CI here :(

**Comment by bam241** [comment]
CI is back... this is strange

**Comment by pshriwise** [comment]
Should be running now.

**Comment by pshriwise** [comment]
I'm thinking the next step here is to update the CI Docker image, make sure it passes, and then merge. 

Does that sound good @bam241 @gonuke?

**Comment by gonuke** [comment]
I've lost the train of thought here a little, but is the current failing test because we don't have a shared object build of G4 on the image?  Why does that only affect U18.04/gcc?

**Comment by pshriwise** [comment]
Correct. That only affects 18.04 for the reasons described in #673.

**Comment by ljacobson64** [comment]
There is a shared object build of Geant4 in the docker image. Even though the [cmake command](https://github.com/svalinn/DAGMC/blob/develop/CI/docker/build_geant4.sh#L19) contains `-DBUILD_STATIC_LIBS=ON` and does not contain anything about building shared libraries, shared libraries are still built by default. I believe the tests would pass if Pat's change [here](https://github.com/svalinn/DAGMC/pull/674/files#diff-53211a5f6056e7acab03b692b7088c3cR24) were incorporated into the docker image.

**Comment by bam241** [comment]
I think, the new docker container should be pushed, to make sure the failing test is passing
and then merge this PR once everything is working well.

Other PRs will probably not be sensitive to those docker changes anyway....

**Comment by pshriwise** [comment]
So shall I go ahead and push the new image via `update_docker.sh` then?

**Comment by bam241** [comment]
> So shall I go ahead and push the new image via `update_docker.sh` then?

I vote yes !

**Comment by pshriwise** [comment]
Shockingly, everything seems to have gone according to plan. Let me know if there are any other adjustments I should make.

**Review by bam241 — APPROVED** [review]
this looks great !

will merge this afternoon unless someone comments.

**Comment by gonuke** [comment]
Thanks @pshriwise

**Pull request merged**
